packages/puppeteer-core/src/api/Locator.ts

```diff
@@ -0,0 +1,374 @@
+   * exceeded. This timeout should be generally much lower as locating an
+   * element means multiple asynchronious operations.
+   */
+  operationTimeout: number;
+}
+
+/**
+ * @internal
+ */
+type ActionCondition = (
+  element: ElementHandle,
+  signal: AbortSignal
+) => Promise<void>;
+
+/**
+ * @internal
+ */
+export interface ActionOptions {
+  signal?: AbortSignal;
+  conditions: ActionCondition[];
+}
+
+/**
+ * All the events that a locator instance may emit.
+ *
+ * @internal
+ */
+export enum LocatorEmittedEvents {
+  /**
+   * Emitted every time before the locator performs an action on the located element(s).
+   */
+  Action = 'action',
+}
+
+/**
+ * @internal
+ */
+export interface LocatorEventObject {
+  [LocatorEmittedEvents.Action]: never;
+}
+
+/**
+ * Locators describe a strategy of locating elements and performing an action on
+ * them. If the action fails because the element are not ready for the action,
+ * the whole operation is retried.
+ *
+ * @internal
+ */
+export class Locator extends EventEmitter {
+  #page: Page;
+  #selector: string;
+  #options: LocatorOptions;
+
+  constructor(
+    page: Page,
+    selector: string,
+    options: LocatorOptions = {
+      visibility: 'visible',
+      timeout: page.getDefaultTimeout(),
+      operationTimeout: 1000,
+    }
+  ) {
+    super();
+    this.#page = page;
+    this.#selector = selector;
+    this.#options = options;
+  }
+
+  override on<K extends keyof LocatorEventObject>(
+    eventName: K,
+    handler: (event: LocatorEventObject[K]) => void
+  ): Locator {
+    return super.on(eventName, handler) as Locator;
+  }
+
+  override once<K extends keyof LocatorEventObject>(
+    eventName: K,
+    handler: (event: LocatorEventObject[K]) => void
+  ): Locator {
+    return super.once(eventName, handler) as Locator;
+  }
+
+  override off<K extends keyof LocatorEventObject>(
+    eventName: K,
+    handler: (event: LocatorEventObject[K]) => void
+  ): Locator {
+    return super.off(eventName, handler) as Locator;
+  }
+
+  /**
+   * Retries the `fn` until a truthy result is returned.
+   */
+  async #waitForFunction(
+    fn: (signal: AbortSignal) => unknown,
+    signal?: AbortSignal,
+    timeout = this.#options.operationTimeout
+  ): Promise<void> {
+    let isActive = true;
+    let isUserAborted = false;
+    let iterationController: AbortController;
+    // If the loop times out, we abort only the last iteration's controller.
+    const timeoutId = setTimeout(() => {
+      isActive = false;
+      iterationController?.abort();
+    }, timeout);
+    // If the user's signal aborts, we abort the last iteration and the loop.
+    signal?.addEventListener(
+      'abort',
+      () => {
+        iterationController?.abort();
+        isUserAborted = true;
+      },
+      {once: true}
+    );
+    while (isActive && !isUserAborted) {
+      iterationController = new AbortController();
+      try {
+        const result = await fn(iterationController.signal);
```

Comment:
What about the user's signal?

Reply:
on line 147, if the user signal is aborted, the iteration controller is aborted too.